### PR TITLE
Buffer initdb.tar.zst to a temporary file before upload

### PR DESCRIPTION
**Pull request by arpad-m**
I'm not really sure it is needed but in https://github.com/neondatabase/neon/pull/5912#pullrequestreview-1749982732 , Christian liked the idea of using files instead of buffering the archive to RAM for the *download* path. This is for the upload path, which is probably a very similar situation.

### CODE REVIEW

File: pageserver/src/import_datadir.rs [hunk 1]
```diff
@@ -725,10 +663,10 @@ pub async fn create_tar_zst(pgdata_path: &Utf8Path) -> Result<Vec<u8>> {
     let mut zstd = builder.into_inner().await?;
     zstd.shutdown().await?;
     let compressed = zstd.into_inner();
-    let compressed_len = compressed.buf.len();
-    const INITDB_TAR_ZST_WARN_LIMIT: usize = 2_000_000;
+    let compressed_len = compressed.metadata().await?.len();
```
Comment (problame) [inline pageserver/src/import_datadir.rs:666]: Would be nice if we could avoid the `.metadata()` call by making the zstd builder return the final size on `.shutdown().await`

File: pageserver/src/import_datadir.rs [hunk 2]
```diff
@@ -705,8 +642,9 @@ pub async fn create_tar_zst(pgdata_path: &Utf8Path) -> Result<Vec<u8>> {
     }
     // Do a sort to get a more consistent listing
     paths.sort_unstable();
+    let tempfile = File::from_std(tempfile()?);
```
Comment (problame) [inline pageserver/src/import_datadir.rs:645]: This creates a file in the default tempfile location, i.e., `/tmp`.tmpfs.
We don't want that. We want all files to sit inside the tenant that they belong to, on the `/storage` partition.
I suggest using the `TEMP_FILE_SUFFIX` like we do for other temp files.
You'll need to do the cleanup manually as follows:
1. the load procedure should to the cleanup (`fn reconcile()` is a good starting point to learn more)
2. unlink an existing file here before creating it, and create with `create_new` to detect race conditions

I'm not a fan of `unlink()` directly `create_new()` , like `tempfile()` does.
The reason is that the inode will still grow on disk but we don't see that usage using user-space tools like `du`.

I'd also like this to be a VirtualFile, but, we don't have a tempfile abstraction for VirtualFile yet afaik. The primary challenge there is that the `create` + unlink

File: pageserver/src/import_datadir.rs [hunk 3]
```diff
@@ -725,10 +663,10 @@ pub async fn create_tar_zst(pgdata_path: &Utf8Path) -> Result<Vec<u8>> {
     let mut zstd = builder.into_inner().await?;
     zstd.shutdown().await?;
     let compressed = zstd.into_inner();
-    let compressed_len = compressed.buf.len();
-    const INITDB_TAR_ZST_WARN_LIMIT: usize = 2_000_000;
+    let compressed_len = compressed.metadata().await?.len();
+    const INITDB_TAR_ZST_WARN_LIMIT: u64 = 2_000_000;
```
Comment (problame) [inline pageserver/src/import_datadir.rs:667]: nit


```suggestion
    const INITDB_TAR_ZST_WARN_LIMIT: u64 = 2 * 1024 * 1024;
```

**Review by problame — DISMISSED** [review]

File: pageserver/src/import_datadir.rs (hunk 2 above)
Comment (arpad-m) [reply pageserver/src/import_datadir.rs:645]: There is no `AsyncRead`/`AsyncWrite` impls for VirtualFile so one would have to re-implement all the copy functions. I understand the need to be inside the fd limitation though.

**Comment by arpad-m** [comment]
This PR has lower priority as it doesn't deliver a feature, and the load in terms of memory is pretty low anyways. Any pageserver equipped with >4 GiB can do thousands of initdb upload requests in parallel.

File: pageserver/src/import_datadir.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/import_datadir.rs:645]: You're misreading my comment. VirtualFile is nice to have I think.
The `tempfile()` thing is the blocker here.

File: pageserver/src/import_datadir.rs (hunk 1 above)
Comment (arpad-m) [reply pageserver/src/import_datadir.rs]: can you give concrete code suggestions? `.shutdown().await` has `Result<()>` as the return value. I don't see a way outside of creating a wrapper for the file that tracks the written offsets, which would involve a lot of repetitive code.

File: pageserver/src/import_datadir.rs [hunk 4]
```diff
@@ -707,7 +653,7 @@ pub async fn create_tar_zst(pgdata_path: &Utf8Path) -> Result<Vec<u8>> {
     // Do a sort to get a more consistent listing
     paths.sort_unstable();
     let zstd = ZstdEncoder::with_quality_and_params(
-        YieldingVec::new(),
+        file,
```
Comment (koivunej) [inline pageserver/src/import_datadir.rs:656]: This is good because tokio co-op will now help with yielding.

File: pageserver/src/tenant.rs [hunk 5]
```diff
@@ -3019,6 +3024,12 @@ impl Tenant {
                     backoff::Cancel::new(CancellationToken::new(), || unreachable!()),
                 )
                 .await?;
+
+                if temp_path.exists() {
+                    tokio::fs::remove_file(&temp_path)
+                        .await
+                        .with_context(|| format!("tempfile removal {temp_path}"))?;
+                }
```
Comment (koivunej) [inline pageserver/src/tenant.rs:3032]: We could avoid this exists check by just removing the file and ignoring NotFound.

File: pageserver/src/tenant/remote_timeline_client/upload.rs [hunk 6]
```diff
@@ -114,16 +113,16 @@ pub(crate) async fn upload_initdb_dir(
...
     tracing::trace!("uploading initdb dir");
 
-    let size = initdb_dir.len();
-    let bytes = tokio::io::BufReader::new(std::io::Cursor::new(initdb_dir));
+    let bytes = tokio::io::BufReader::new(initdb_tar_zst);
 
     let remote_path = remote_initdb_archive_path(tenant_id, timeline_id);
     storage
-        .upload_storage_object(bytes, size, &remote_path)
+        .upload_storage_object(bytes, size as usize, &remote_path)
```
Comment (koivunej) [inline pageserver/src/tenant/remote_timeline_client/upload.rs:125]: I will conflict this with #5446.

Comment (arpad-m) [reply pageserver/src/tenant/remote_timeline_client/upload.rs]: I have merged latest main now and resolved the conflicts

File: pageserver/src/tenant/remote_timeline_client/upload.rs [hunk 7]
```diff
@@ -120,17 +119,17 @@ pub(crate) async fn upload_initdb_dir(
...
+    initdb_tar_zst: File,
+    size: u64,
 ) -> anyhow::Result<()> {
     tracing::trace!("uploading initdb dir");
 
-    let size = initdb_dir.len();
-
-    let bytes = futures::stream::once(futures::future::ready(Ok(initdb_dir)));
+    let buf_reader = tokio::io::BufReader::new(initdb_tar_zst);
+    let file = tokio_util::io::ReaderStream::with_capacity(buf_reader, 8 * 1024);
```
Comment (koivunej) [inline pageserver/src/tenant/remote_timeline_client/upload.rs:128]:
```suggestion
    let file = tokio_util::io::ReaderStream::with_capacity(buf_reader, 32 * 1024);
```

Let's do it on this PR which introduces the file io.

Comment (arpad-m) [reply pageserver/src/tenant/remote_timeline_client/upload.rs]: done

File: pageserver/src/tenant.rs [hunk 8]
```diff
@@ -2988,26 +2988,38 @@ impl Tenant {
...
+                .or_else(|e| {
+                    if e.kind() == std::io::ErrorKind::NotFound {
+                        // If something else already removed the file, ignore the error
+                        Ok(())
+                    } else {
+                        Err(e)
+                    }
+                })
+                .with_context(|| format!("tempfile removal {initdb_tar_zst_path}"))?;
+            if initdb_tar_zst_path.exists() {}
```
Comment (koivunej) [inline pageserver/src/tenant.rs:3002]:
```suggestion
```

File: pageserver/src/tenant.rs [hunk 9]
```diff
@@ -2988,26 +2988,38 @@ impl Tenant {
...
             // Init temporarily repo to get bootstrap data, this creates a directory in the `initdb_path` path
             run_initdb(self.conf, &pgdata_path, pg_version, &self.cancel).await?;
 
             // Upload the created data dir to S3
             if let Some(storage) = &self.remote_storage {
-                let pgdata_zstd = import_datadir::create_tar_zst(&pgdata_path).await?;
-                let pgdata_zstd = Bytes::from(pgdata_zstd);
+                let temp_path = timelines_path.join(format!(
+                    "{INITDB_PATH}.upload-{timeline_id}.{TEMP_FILE_SUFFIX}"
+                ));
```
Comment (koivunej) [inline pageserver/src/tenant.rs:3011]: Using a static tempfile is fine because we are guarded by the mutual exclusion on create_timeline.

**Review by koivunej — APPROVED** [review]

File: pageserver/src/import_datadir.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/import_datadir.rs]: camino_tempfile does also have a builder, but we can go with a static tempfile here as we are finally guarded by a mutex! :)

File: pageserver/src/tenant.rs [hunk 10]
```diff
@@ -2988,26 +2988,38 @@ impl Tenant {
                 .await
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant.rs:2988]: Why is the reader pinned btw?

[the remark this answers is not in the document]
Comment (arpad-m) [reply pageserver/src/tenant.rs:2988]: Regarding the buffer sizes, I've increased them now.

Regarding the Pin, there was a compilation error in some earlier version. now you can remove it and there is no compilation error any more, so I removed it. Good catch!

File: pageserver/src/import_datadir.rs (hunk 2 above)
Comment (arpad-m) [reply pageserver/src/import_datadir.rs]: I have changed it to use a file from the timelines dir, so I'm resolving this.

File: pageserver/src/import_datadir.rs (hunk 1 above)
Comment (arpad-m) [reply pageserver/src/import_datadir.rs]: resolving because I remember dimly that you said that the metadata call was needed.

**Pull request merged**